From ae8ee2d5052816de8c75c80bdf1c2d546ac7c906 Mon Sep 17 00:00:00 2001
From: Bruno Martins <bgcngm@gmail.com>
Date: Thu, 31 Jan 2019 16:09:49 +0000
Subject: [PATCH 2/7] lineage: Properly write rules for Lineage LiveDisplay as
 a HAL

 * This is how Google defines rules for a new HAL, split between
   private (for platform-only types, permissions, and attributes),
   public (for types and attributes) and finally vendor.

Change-Id: I94a299f03bd413afd76ae02a2d979f37eb444709
---
 common/private/hwservice_contexts                |  1 +
 common/private/system_server.te                  |  3 +++
 common/public/attributes                         |  2 ++
 common/public/hal_lineage_livedisplay.te         |  5 +++++
 common/public/hwservice.te                       |  1 +
 common/vendor/file_contexts                      |  4 ++--
 common/vendor/hal_lineage_livedisplay_default.te |  8 ++++++++
 common/vendor/hal_livedisplay_default.te         | 12 ------------
 common/vendor/hwservice.te                       |  2 --
 common/vendor/hwservice_contexts                 |  2 --
 common/vendor/system_server.te                   |  3 ---
 qcom/common/hal_lineage_livedisplay_default.te   | 14 ++++++++++++++
 qcom/common/hal_livedisplay_default.te           | 14 --------------
 13 files changed, 36 insertions(+), 35 deletions(-)
 create mode 100644 common/private/hwservice_contexts
 create mode 100644 common/public/attributes
 create mode 100644 common/public/hal_lineage_livedisplay.te
 create mode 100644 common/public/hwservice.te
 create mode 100644 common/vendor/hal_lineage_livedisplay_default.te
 delete mode 100644 common/vendor/hal_livedisplay_default.te
 create mode 100644 qcom/common/hal_lineage_livedisplay_default.te
 delete mode 100644 qcom/common/hal_livedisplay_default.te

diff --git a/common/private/hwservice_contexts b/common/private/hwservice_contexts
new file mode 100644
index 0000000..2c5d9ce
--- /dev/null
+++ b/common/private/hwservice_contexts
@@ -0,0 +1 @@
+vendor.lineage.livedisplay::IColor                   u:object_r:hal_lineage_livedisplay_hwservice:s0
diff --git a/common/private/system_server.te b/common/private/system_server.te
index 82e25fb..9d7a8c6 100644
--- a/common/private/system_server.te
+++ b/common/private/system_server.te
@@ -2,3 +2,6 @@ allow system_server storage_stub_file:dir getattr;
 
 # Allow LineageHW (running as system server) to access LiveDisplay tuneables
 allow system_server sysfs_livedisplay_tuneable:file rw_file_perms;
+
+# Use HALs
+hal_client_domain(system_server, hal_lineage_livedisplay)
diff --git a/common/public/attributes b/common/public/attributes
new file mode 100644
index 0000000..785e7dd
--- /dev/null
+++ b/common/public/attributes
@@ -0,0 +1,2 @@
+# HALs
+hal_attribute(lineage_livedisplay)
diff --git a/common/public/hal_lineage_livedisplay.te b/common/public/hal_lineage_livedisplay.te
new file mode 100644
index 0000000..6a9aac5
--- /dev/null
+++ b/common/public/hal_lineage_livedisplay.te
@@ -0,0 +1,5 @@
+# HwBinder IPC from client to server
+binder_call(hal_lineage_livedisplay_client, hal_lineage_livedisplay_server)
+
+add_hwservice(hal_lineage_livedisplay_server, hal_lineage_livedisplay_hwservice)
+allow hal_lineage_livedisplay_client hal_lineage_livedisplay_hwservice:hwservice_manager find;
diff --git a/common/public/hwservice.te b/common/public/hwservice.te
new file mode 100644
index 0000000..fb4890c
--- /dev/null
+++ b/common/public/hwservice.te
@@ -0,0 +1 @@
+type hal_lineage_livedisplay_hwservice, hwservice_manager_type;
diff --git a/common/vendor/file_contexts b/common/vendor/file_contexts
index 17704d2..dcca85a 100644
--- a/common/vendor/file_contexts
+++ b/common/vendor/file_contexts
@@ -5,8 +5,8 @@
 /(vendor|system/vendor)/bin/hw/android\.hardware\.light@2\.0-service\.aw2013 u:object_r:hal_light_default_exec:s0
 
 # LiveDisplay HAL
-/(vendor|system/vendor)/bin/hw/vendor\.lineage\.livedisplay@1\.0-service-legacymm u:object_r:hal_livedisplay_default_exec:s0
-/(vendor|system/vendor)/bin/hw/vendor\.lineage\.livedisplay@1\.0-service-sdm u:object_r:hal_livedisplay_default_exec:s0
+/(vendor|system/vendor)/bin/hw/vendor\.lineage\.livedisplay@1\.0-service-legacymm u:object_r:hal_lineage_livedisplay_default_exec:s0
+/(vendor|system/vendor)/bin/hw/vendor\.lineage\.livedisplay@1\.0-service-sdm u:object_r:hal_lineage_livedisplay_default_exec:s0
 
 # USB HAL
 /(vendor|system/vendor)/bin/hw/android\.hardware\.usb@1\.0-service\.basic u:object_r:hal_usb_default_exec:s0
diff --git a/common/vendor/hal_lineage_livedisplay_default.te b/common/vendor/hal_lineage_livedisplay_default.te
new file mode 100644
index 0000000..7a85640
--- /dev/null
+++ b/common/vendor/hal_lineage_livedisplay_default.te
@@ -0,0 +1,8 @@
+type hal_lineage_livedisplay_default, domain;
+hal_server_domain(hal_lineage_livedisplay_default, hal_lineage_livedisplay)
+
+type hal_lineage_livedisplay_default_exec, exec_type, vendor_file_type, file_type;
+init_daemon_domain(hal_lineage_livedisplay_default)
+
+# Allow LiveDisplay HAL's default implementation to use vendor-binder service
+vndbinder_use(hal_lineage_livedisplay_default)
diff --git a/common/vendor/hal_livedisplay_default.te b/common/vendor/hal_livedisplay_default.te
deleted file mode 100644
index 874d446..0000000
--- a/common/vendor/hal_livedisplay_default.te
+++ /dev/null
@@ -1,12 +0,0 @@
-type hal_livedisplay_default, domain;
-hwbinder_use(hal_livedisplay_default)
-vndbinder_use(hal_livedisplay_default)
-
-type hal_livedisplay_default_exec, exec_type, vendor_file_type, file_type;
-init_daemon_domain(hal_livedisplay_default)
-
-# Check if hwservicemanager is ready
-get_prop(hal_livedisplay_default, hwservicemanager_prop)
-
-# Add vendor.lineage.livedisplay::IColor service to hwservicemanager
-add_hwservice(hal_livedisplay_default, hal_lineage_livedisplay_hwservice)
diff --git a/common/vendor/hwservice.te b/common/vendor/hwservice.te
index 8f1a791..2ca8189 100644
--- a/common/vendor/hwservice.te
+++ b/common/vendor/hwservice.te
@@ -1,3 +1 @@
-type hal_lineage_livedisplay_hwservice, hwservice_manager_type;
-
 type hal_lineage_power_hwservice, hwservice_manager_type;
diff --git a/common/vendor/hwservice_contexts b/common/vendor/hwservice_contexts
index 193b712..3ae7fd7 100644
--- a/common/vendor/hwservice_contexts
+++ b/common/vendor/hwservice_contexts
@@ -1,3 +1 @@
-vendor.lineage.livedisplay::IColor       u:object_r:hal_lineage_livedisplay_hwservice:s0
-
 vendor.lineage.power::ILineagePower      u:object_r:hal_lineage_power_hwservice:s0
diff --git a/common/vendor/system_server.te b/common/vendor/system_server.te
index 942a293..4a889fc 100644
--- a/common/vendor/system_server.te
+++ b/common/vendor/system_server.te
@@ -1,4 +1 @@
-allow system_server hal_lineage_livedisplay_hwservice:hwservice_manager find;
-binder_call(system_server, hal_livedisplay_default)
-
 allow system_server hal_lineage_power_hwservice:hwservice_manager find;
diff --git a/qcom/common/hal_lineage_livedisplay_default.te b/qcom/common/hal_lineage_livedisplay_default.te
new file mode 100644
index 0000000..5ca366c
--- /dev/null
+++ b/qcom/common/hal_lineage_livedisplay_default.te
@@ -0,0 +1,14 @@
+# Do not use add_service() as hal_graphics_composer_default may be the provider as well
+allow hal_lineage_livedisplay_default qdisplay_service:service_manager find;
+
+binder_call(hal_lineage_livedisplay_default, hal_graphics_composer_default)
+
+# Allow LiveDisplay to store files under /data/vendor/display and access them
+allow hal_lineage_livedisplay_default display_vendor_data_file:dir rw_dir_perms;
+allow hal_lineage_livedisplay_default display_vendor_data_file:file create_file_perms;
+
+# Allow LiveDisplay to access vendor display property
+get_prop(hal_lineage_livedisplay_default, vendor_display_prop)
+
+# Allow LiveDisplay to access pps socket
+unix_socket_connect(hal_lineage_livedisplay_default, pps, mm-pp-daemon)
diff --git a/qcom/common/hal_livedisplay_default.te b/qcom/common/hal_livedisplay_default.te
deleted file mode 100644
index 24a65b1..0000000
--- a/qcom/common/hal_livedisplay_default.te
+++ /dev/null
@@ -1,14 +0,0 @@
-# Do not use add_service() as hal_graphics_composer_default may be the provider as well
-allow hal_livedisplay_default qdisplay_service:service_manager find;
-
-binder_call(hal_livedisplay_default, hal_graphics_composer_default)
-
-# Allow LiveDisplay to store files under /data/vendor/display and access them
-allow hal_livedisplay_default display_vendor_data_file:dir rw_dir_perms;
-allow hal_livedisplay_default display_vendor_data_file:file create_file_perms;
-
-# Allow LiveDisplay to access vendor display property
-get_prop(hal_livedisplay_default, vendor_display_prop)
-
-# Allow LiveDisplay to access pps socket
-unix_socket_connect(hal_livedisplay_default, pps, mm-pp-daemon)
-- 
2.17.1

